common/src/main/java/io/netty/util/ByteString.java

```diff
@@ -317,36 +318,42 @@ public final int forEachByteDesc(ByteProcessor visitor) throws Exception {
      *         The last-visited index If the {@link ByteProcessor#process(byte)} returned {@code false}.
      */
     public final int forEachByteDesc(int index, int length, ByteProcessor visitor) throws Exception {
-        if (index < 0 || length > value.length - index) {
+        if (index < 0 || length > length() - index) {
             throw new IndexOutOfBoundsException("expected: " + "0 <= index(" + index + ") <= start + length(" + length
-                    + ") <= " + "length(" + value.length + ')');
+                    + ") <= " + "length(" + length() + ')');
         }
 
-        for (int i = index + length - 1; i >= index; --i) {
+        final int end = offset + index;
+        for (int i = offset + index + length - 1; i >= end; --i) {
             if (!visitor.process(value[i])) {
-                return i;
+                return i - offset;
             }
         }
         return -1;
     }
 
     public final byte byteAt(int index) {
-        return value[index];
+        // We must do a range check here to enforce the access does not go outside our sub region of the array.
+        // We rely on the array access itself to pick up the array out of bounds conditions
+        if (index < 0 || index >= length) {
```

Comment:
Can you be more explicit (do you mean `if (index - length > 0) { /*ERROR*/} `)? I'm not sure this completely covers all cases.  I need to make sure the index is in the range `[0, length)`.  However it seems like there are always boundary conditions where the integer arithmetic may wrap and give incorrect results.